### ISSUE TITLE
Implementation details

### ISSUE THREAD

**Issue by pickten** [pickten/Conquest#1]
Let's use this issue to talk implementation & git setup. For now, players and graphics is irrelevant, so to hell with that.

I've created a branch `pieces` for the work I'm doing on pieces; @PlatyPrograms should do similar for `map`; use `git branch map; git checkout map; git pull origin map; [WORK]; git push origin map` to use it (as opposed to `git pull origin` and `git push origin`), and the github to view (or you could switch back & forth but github is good/easier for visualization). We can merge the branches to master when they are closer done. Also, btw, you should click "Watch" at the top right of the page to watch the repo and get notifs of most changes.

In terms of implementation details, `C` lacks generics, so we'll not have much overlap AFAIK.

SO FAR, I'm assuming the existence of a generic `nodeType` to match `pieceType` that I'm using. Should be trivial to add.

Loosely, `piece`s have moves left and what they carry; `pieceType`s have max allowed moves and a bunch of `pieceAction`s and `nodeAction`s that represent whether they can by default do a particular action (e.g. travel to a given `nodeType`, carry a given `pieceType`), and all relevant exceptions (currently as an array, but soon an RB tree). Also some `intPieceAction`s, which have extra int data for each.

I'm going to guess the map will be similarly `node`s containing structural and game info, each containing a `nodeType` representing land/sea/space/other types. Ports will actually be simple, as they are just `Land` nodes linked to `Water` nodes.

Up to now, things should be reasonable, but there's two crucial things left: getting all this stuff from a file (sidenote: Haskell's parsing is great, and FFI should be simple to set up, so that's viable if this turns out to be complex) and figuring out what the allowable actions should be.

Here's what I have for actions ATM. Is anything missing or whatnot? One that ocurred to me is multimount (mounting a mounted unit). Worth adding?

| action | node/piece | notes |
| ------- |-------------- | ------ |
| kill | piece | kill the piece |
| carry | int+piece | carry the piece (max & weight of each piece)\* |
| combo | piece | max moves resets when killing this piece modulo recapture |
| recapture | int+piece | when killing this, can be recaptured (and delay) |
| travel | node | travel to the node |
| mountKill | piece | can kill the piece and mount it (e.g. land vs ship capturing in Conquest) |
| dismountKill | piece | can kill the piece when dismounting |

\* I could probably split this up into an array of "carry types" each with `intPieceAction`s, but that's probably stupid

**Comment by pickten** [comment]
Some random data structure talk:
---

So I'm working on three general data structure files (`rbtrees`, `stack`, and `linkedlist`) which will be generic (handled via macros) implementations of red-black trees, stacks, and linked lists; the former requires `struct RBTREE_TYPE` to contain an `int id`, the latter two a `struct STACK_TYPE` and `struct LINKED_TYPE` with no restrictions (both default to a placeholder with `int id` if needbe and `void* item`).

`stack` and `linkedlist` are pretty much finished, so ~~I'll commit them to `pieces` soon~~ they're committed; `rbtrees` is pretty messy and incompleteso not for now; I'm steadily going through the wikipedia implementation to see what I can adapt. It's possible I'll have to give nodes  `parent` pointers, but I'd really rather avoid it and use a stack-based implementation because it makes the trees a bit simpler to create. This *should* be doable, but I've also not used rb-trees so who knows.

One pain point worth watching out for is that they have guards against multi-inclusion for each type input, but the relevant flags need to be changed at the call site. Thankfully, `stack` knows that it's called in `rbtrees` generically, and will apply the flag in `rbtrees` and likewise for `linkedlist` and `stack` and `rbtrees`. There may be a better approach, but I don't know of any ATM. In any case, we can feel free to rely on them but know that we'll have to do, e.g. `#define LINKED_H_player 1` when we set up the `player` linked list. Basically, what's going on is that the usual `#ifndef ... #define ... #endif` pattern fails since you'd need to expand macros before checking `#ifndef` etc, but this is impossible. So instead, there's a bit of jankiness and we use `#if NDEF(AFFIX(...)) ... #endif`, where `NDEF` and `AFFIX` are some custom macros (in a file `genericmacros` that I'll also commit) exclusively useful for this. This works since `#if` takes any expression it can evaluate at compile time (thus allowing macros), but also means you can't rely on `defined` (which is how `#ifndef` works) and need a custom way to tell if stuff is defined. Now, apparently if the flag isn't defined it'll evaluate to `0`, so `NDEF` is just checks `== 0`.

In summary, to hell with generics, macros and especially `#ifndef`; to create data structures and core functions for a type, you need to do, e.g. 

```c
#ifndef LINKED_H_player /* this is still ok since it's not generic */
#define LINKED_TYPE player
#include "linkedlist.h"
#undef LINKED_TYPE
#define LINKED_H_player 1
#endif
```

and no, I literally cannot make a macro to do that.

**Comment by PlatyPrograms** [comment]
Hey here's a problem that might be worth thinking about

Do we implement moves on a 1-by-1 basis (i.e. you literally need to enter each move command) or a full-move basis (i.e. you can tell a piece to move three spaces and it does that)?

It seems to me we would encounter a whole bunch of overhead trying to implement the latter because we would need to find an optimal route, but the former is really user-unfriendly

Also i'm going to give each map node a pointer to the piece currently on it, might be useful

**Comment by PlatyPrograms** [comment]
Another thing:

It might be easier to write an initialization function for each map (there are only  4 of them) than screw around with file parsing

Just a thought, in the meantime I'm just going to assume we can figure out the file parsing and write code based on that

**Comment by pickten** [comment]
I've been assuming moves are 1-by-1. A full-move basis would probably not be hard, but would essentially require 1-by-1 so there's no point in not doing 1-by-1 to start. Regardless, because of recapturing, the moving system should not try to move after a kill, which makes things much simpler (treat enemies as obstacles, then depth first and run whatever comes out of that one step at a time) and fits better with the recapture rules (since you can't kill on the way). Besides, I've not yet finished RB-trees and I should do that first.

Regarding parsing: That's what I've been doing as well. Assuming the config format is sane, parsing should be really easy. Don't know of any good `C` parsing resources off hand, but it should not be hard; worst comes to worst I write a Haskell parser (I've done graph parsing before, which is probably the most representative non-FFI portion, and it's really basic).

**Comment by PlatyPrograms** [comment]
I've worked out a basic format for the source file of a map:

`node0Cnode1node2node4`
`node1Lnode2`
`...`
`node4Pnode5node6`
`node5Snode7`

Essentially this means:
`node0` is always the white capitol; black capitol can change.
`node0` is of type Capitol, and is adjacent to nodes 1, 2, and 4.
 Each line of config takes the form `NameTypeAdj1Adj2...\n`, meaning parsing of it is as simple as converting to an array of chars and iterating through it.

Note: `adj0` is always defined as the node from which the current node was initialized, so it doesn't need to be explicitly stated.

I'm writing the rest of my code assuming this

**Comment by PlatyPrograms** [comment]
And another update

It turns out that C doesn't actually allow for runtime variable creation (which means no runtime node creation), which means that we're going to have to write the initialization code for each map and then give an input option to specify which map to load.

The good news is that creating the maps are pretty simple, since they're just a bunch of linked lists.

For the purpose of testing I'm going to implement a super-simple map with two 1-space capitols, one port, and some dinky land.

Also I'm ditching the whole "nodes generating nodes" idea in favor of a simpler system whereby each node just has the pointers to the nodes next to it and are defined in a function.

**Comment by pickten** [comment]
Finished rb insertion; probably needs to be refactored (140 LOC) but meh.
Still need to do deletion and maybe validation (i.e. checking that a tree meets the rb-tree properties), and then I'll get back to directly necessary stuff. 

Haven't thought much about the `piece` format, but probably a something like `\n` denotes `pieceType` end, and each contains `;` separated data, each with `,` separated fields. Each entry has one field to indicate what is being filled in (since there will be plenty of defaults), and then a series of ints denoting max (in the case of `travel`, `recapture` or `carry`), default, and then special case ids. It will probably be worth writing a comand-line config manager, probably in something simple like Python that has good support for hash maps. 

Also, we should totally be able to do runtime node creation; I'm not sure what you're thinking of. You can't dynamically change how many variables you have (very few languages can without a hash map), but we don't need that to build a graph. Regardless, you can just write everything in terms of a `getMap` function that returns the config (either by parsing or hardcoded), and we can draft a config spec and parser later, so that the main logic is independent of how it gets the config. I'll be doing the same for `pieces`.

**Comment by pickten** [comment]
Pushed the rest of rb tree stuff (validation will come later because I'm lazy and don't technically need it). Now back to `pieces` I go.

EDIT: What will I be doing to go from a `Node` to whatever's on it? Something like `node->piece`? Also, btw, I'm going to put movement into `player.c` in a separate branch.

EDIT 2: Decided to hold off on actual movement and stuff until the `Node` type itself is more finalized.

**Comment by PlatyPrograms** [comment]
Player.c is going to be real basic--just a move count and a 1/0 turn, governed by the play loop in main.c.

As far as movement, I think the implementation is something like

    // assume piece, destination given

    source = piece->node;

    toMove = piece;

    if ((destination->piece == NULL or destination->piece->embarkable == 1 or (destination->piece->team == player->opponent and destination->piece->type == capturable) and piece->moves != 0 ) {
        source->piece = NULL;

        destination->piece = piece;

        piece->moves -= 1;

    return 0;


Obviously 'embarkable' and 'capturable' are placeholders for more nuanced implementations of 'can this piece embark/capture that piece'

**Comment by pickten** [comment]
Yeah, it is pretty straightforwards. I was just planning on getting it done now because `pieces` is pretty much limited to a `struct` declaration (sidenote: I haven't `typedef`'d anything; but this is done), creation/deletion (didn't think to do this earlier, for some reason), `canTravel`/`canKill`/`canCarry`/etc. (handled via macro and done) and parsing.

**Comment by PlatyPrograms** [comment]
SUPER IMPORTANT

DO NOT USE THE VARIABLE `nodes` FOR ANYTHING

IT IS DECLARED AS AN EXTERN IN `mapinit.h` AND HOLDS THE ADDRESSES OF ALL THE NODES

**Comment by pickten** [comment]
Changed all uses of `nodes`, thanks for the heads up. It now uses `ancestorStack` which should be fine.

Besides that, my variables are either obvious and postfixed with type, `node`, `parent`, `grandparent`, `uncle`, and `sibling`. Let me know if those need changing.

Oh btw: I'm using the convention that the constructor of type `t` is `newT`, destructor `deleteT`, and default constructor (non-extant if minimal requires as much info as the full constructor does)  `emptyT`, if you were wondering.

**Comment by PlatyPrograms** [comment]
Also btw the individual nodes are named `node0`-`noden`, so if you are referencing them that might be useful information. 

There used to be `head, node1`-`noden`, but that was confusing so now `node0` acts as `head`

**Comment by pickten** [comment]
Aren't those local to the `testmapinit` function?

Also, line `43-44` of `mapinit.c` (the declaration of `nodes`) seems bugged and should be using `{}` instead of `()` (`(a,b,c)` means "evaluate `a`, `b`, `c`, etc. and return the last"; you want `{a, b, c}`) unless I'm missing something

**Comment by PlatyPrograms** [comment]
For moving, I think the pieces need to know what node they're on and the nodes need to know what pieces they have on them, so the nodes will eventually be globals; I just haven't gotten around to declaring them as such because for the purpose of getting a working testmapinit it hasn't been necessary

And yes, that part of the code is bugged as shit and I'm working on fixing it

**Comment by pickten** [comment]
I disagree that the nodes should be global. Ignoring the fact that that kills any hope of writing a parser, even if the piece needs a pointer to it's parent node (and I don't think that's true, give me a minute), that doesn't make the nodes global -- at worst it would mean searching through the nodes for the holder, and that's best done with a (perhaps global) BST/RB tree (I'd need to modify the RB tree code for this, but it's an easy change). More practically, we just cave and add pointers both ways, while keeping everything out of the global scope.

Also, as for why I don't think we need to go both ways: the `move` function is simplest if we pass the input and output nodes and carrier into the function directly. I mean, think about the end goal for a sec: before the user even selects the piece to move, they will have to input the node its on and we'll need to locate it in the carrying hierarchy, so we will have know the carrier and node it's on before we can even tell the piece to move!

**Comment by pickten** [comment]
Oh, btw, a minor time saver macro'd way to write the tedious chunk of `mapinit`

```c
#define DOFOR(x) do {x(1);x(2);x(3);x(4);x(5);x(6);x(7);x(8);x(9);x(10);x(11);x(12);x(13);x(14);} while(0)
#define INIT_NODE(i) Node * node ## i
#define INTO_NODES(i) nodes[i] = node ## i
DOFOR(INIT_NODE);
nodes[0] = node0;
DOFOR(INTO_NODES);
#undef DOFOR
#undef INIT_NODE
#undef INTO_NODES
```

**Comment by PlatyPrograms** [comment]
At the very least `nodes` will need to be global or at least shared with main so that the function can check the legality of a move. `pieces` not so as long as we make sure the nodes also know what piece is on them (it's easier to check the target node's current piece than to search all the pieces for one that is on that node)

**Comment by pickten** [comment]
Not sure I follow what you're saying. Doesn't move legality come down to info that needs to be passed in (or found) regardless? (e.g. the target `node` and its type, the `piece`'s data, `piece` ownership). 

In any case I'm basically done with the stuff that's just for `pieces` so I'll set up a PR, and will move on to config parsing, followed by a separate config utility (I know, neither of these are strictly needed but I want to write them), so if you have a sense of how to handle  `player` stuff (including movement/etc.) go ahead.

**Comment by PlatyPrograms** [comment]
For a move to be legal, it must:

-be to an adjacent node 

-Contain no pieces or a capturable enemy piece or an embarkable friendly piece 

-be a legal type(e.g. Not sea)

If the nodes are global and know what's on them, all of these checks are stupidly easy to perform and we just need to pass the function the destination node, target node, and piece to move 

Hence nodes should be global

**Comment by pickten** [comment]
Oh, by the way, can you give `Node`s an `int id` field? I need it for parsing (I mean, I could use a wrapper struct, but it's simpler this way). Other things I use that don't exist are a `Node` creator `newNode(int id)`, destructor `deleteNode(Node *n)`, and helper funcs `attachPieceToNode`, `getPieceFromNode`, and `addLinkToNode` (that adds both links).

In any case the spec (it's implemented!) looks like:

 - `;` and empty line: ignored, for comments and spacing

 - `(...)`: node, with one line for piece type on it (-1 for none), one line with the node type and one line for each id of a linked node. The piece type *must* be declared before any node that uses it (using nodes ahead of time is fine, as are piece types in a piece action). This should not be a problem if you do your piece types first, which is what I'd recommend regardless, so I'm not going to change it to allow using piece types before they're defined in nodes.

 - `{...}`: pieceType, with a couple lines of int prelude (`maxMoves`, `carryCap`), then split up into chunks by letters denoting what the action being defined is (the first letter of the name, lowercase, except for combo which is uppercase to avoid conflicting with carry). For actions that involve extra ints, put the `id` on one line and the int on the next. Node types (for travel) should be entered with the type as an ASCII value, not as a char because it's simpler this way to implement. The default state will be parsed as `false`; to set it to true, use a second copy of the letter. e.g. `{...kk \n c...}` will be able to kill anything.

 - `#...`: include the file named after `#`. You would not believe the stupid shit in the implementation that stems from this.

Indentation besides ` ` and `\t` is invalid. 
Trailing spaces may or may not also be invalid, depending on whether or not `atoi` (I know it sucks, but meh) can handle them.

The implementation is pretty terrible and probably needs heavy refactoring (I used some macro tricks, for instance).

edit: realized I forgot to take ownership into account. Doy. Will fix in a minute

edit 2: Also realized I forgot to add naming. Bleh. Fixed now (name goes right after `{` before the first newline)

edit 3: haven't done ownership b/c I dunno how to give players ownership of a piece. Currently, I'm relying on a func `givePlayerPiece(int playerID, piece)` but I'm guessing that's not all that workable.

**Comment by pickten** [comment]
Again, not sure the fact that the node is global comes into play at all. `n` adjacent to `m` comes down to `n` having a pointer to `m`, the enemy pieces thing comes down to either giving pieces pointers to their owners or vice versa (also not sure of any other way to do it), and legal types are already stored on the pieces themselves (since they hold pointers to their types, which stores all the piece's rules).

 (I certainly agree nodes should know what's on them, though, didn't mean to sound like I was disagreeing with that)

**Comment by PlatyPrograms** [comment]
I can come up with the node building functions pretty easily 

As far as globality goes, I'll implement the players and deal with that the way I have worked out so the point is sort of moot

**Comment by PlatyPrograms** [comment]
Let's get the ball rolling on this again

With a whole break to work with, we should be able to finish up what we have and merge it into master by January
